pkg/cmd/experimental/syncgroups/ad/ldapinterface_test.go

```diff
@@ -0,0 +1,206 @@
+		if len(testCase.cacheSeed) > 0 {
+			ldapInterface.cachedGroupedUsers = testCase.cacheSeed
+		}
+		members, err := ldapInterface.ExtractMembers(testGroupUID)
+		if !reflect.DeepEqual(err, testCase.expectedError) {
+			t.Errorf("%s: incorrect error returned:\n\texpected:\n\t%v\n\tgot:\n\t%v\n", testCase.name, testCase.expectedError, err)
+		}
+		if !reflect.DeepEqual(members, testCase.expectedMembers) {
+			t.Errorf("%s: incorrect members returned:\n\texpected:\n\t%v\n\tgot:\n\t%v\n", testCase.name, testCase.expectedMembers, members)
+		}
+	}
+}
+
+func TestListGroups(t *testing.T) {
+	client := testclient.NewDNMappingClient(
+		testclient.New(),
+		map[string]*ldap.Entry{
+			usersBaseDN: testUserEntry,
+		},
+	)
+	ldapInterface := newTestADLDAPInterface(client)
+	groups, err := ldapInterface.ListGroups()
+	if !reflect.DeepEqual(err, nil) {
+		t.Errorf("listing groups: incorrect error returned:\n\texpected:\n\t%v\n\tgot:\n\t%v\n", nil, err)
+	}
+	if !reflect.DeepEqual(groups, []string{testGroupUID}) {
+		t.Errorf("listing groups: incorrect group list:\n\texpected:\n\t%v\n\tgot:\n\t%v\n", []string{testGroupUID}, groups)
+	}
+}
+
+func TestPopulateCache(t *testing.T) {
+	var testCases = []struct {
+		name             string
+		cacheSeed        map[string][]*ldap.Entry
+		searchDNOverride string
+		client           ldapclient.Client
+		expectedError    error
+		expectedCache    map[string][]*ldap.Entry
+	}{
+		{
+			name: "cache already populated",
+			cacheSeed: map[string][]*ldap.Entry{
+				testGroupUID: {testUserEntry},
+			},
+			expectedError: nil,
+			expectedCache: map[string][]*ldap.Entry{
+				testGroupUID: {testUserEntry},
+			},
+		},
+		{
+			name:          "user query error",
+			client:        testclient.NewMatchingSearchErrorClient(testclient.New(), usersBaseDN, searchErr),
+			expectedError: searchErr,
+			expectedCache: make(map[string][]*ldap.Entry), // won't be nil but will be empty
+		},
+		{
+			name: "cache populated correctly",
+			client: testclient.NewDNMappingClient(
+				testclient.New(),
+				map[string]*ldap.Entry{
+					usersBaseDN: testUserEntry,
+				},
+			),
+			expectedError: nil,
+			expectedCache: map[string][]*ldap.Entry{
+				testGroupUID: {testUserEntry},
+			},
+		},
+	}
+	for _, testCase := range testCases {
+		ldapInterface := newTestADLDAPInterface(testCase.client)
+		if len(testCase.cacheSeed) > 0 {
+			ldapInterface.cachedGroupedUsers = testCase.cacheSeed
+		}
+		err := ldapInterface.populateCache()
+		if !reflect.DeepEqual(err, testCase.expectedError) {
+			t.Errorf("%s: incorrect error returned:\n\texpected:\n\t%v\n\tgot:\n\t%v\n", testCase.name, testCase.expectedError, err)
+		}
+		if !reflect.DeepEqual(ldapInterface.cachedGroupedUsers, testCase.expectedCache) {
+			t.Errorf("%s: incorrect cache state:\n\texpected:\n\t%v\n\tgot:\n\t%v\n", testCase.name, testCase.expectedCache, ldapInterface.cachedGroupedUsers)
+		}
+	}
+}
+
+var searchErr error = errors.New("generic search error")
+
+const (
```

Comment:
Also, I'm not a fan, but I don't care enough to change it.  This is forcing too much structure on the tests because different units are starting to rely on these and they aren't logically required.